capsules/src/console.rs

```diff
@@ -162,6 +175,39 @@ impl<'a, U: UART> Console<'a, U> {
             app.write_buffer = Some(slice);
         }
     }
+
+    /// Internal helper function for starting a receive operation
+    fn receive_new(&self, app_id: AppId, app: &mut App, len: usize) -> ReturnCode {
+        if self.rx_buffer.is_none() {
+            // For now, we tolerate only one concurrent receive operation on this console.
+            // Competing apps will have to retry until success.
+            return ReturnCode::EBUSY;
+        }
+
+        match app.read_buffer {
+            Some(ref slice) => {
+                let read_len = cmp::min(len, slice.len());
+                if read_len > self.rx_buffer.map_or(0, |buf| buf.len()) {
+                    // For simplicity, impose a small maximum receive length
+                    // instead of doing incremental reads
+                    ReturnCode::EINVAL
+                }
+                else {
+                    // Note: We have ensured above that rx_buffer is present
+                    app.read_len = read_len;
+                    self.rx_buffer.take().map(|buffer| {
+                        self.rx_in_progress.set(Some(app_id));
+                        self.uart.receive(buffer, app.read_len);
```

Comment:
why not `self.uart,receive(&buffer[0..app.read_len])`? Passing the length explicitly seems unnecessary to me!

Reply:
to receive _into_ that buffer, it has to be mutable. for it to be mutable, it _has_ to be returned to the caller through the callback. if we pass in a sliced buffer, there will be no way to recover the complete slice in the callback.

So far, the solution we've been using is to pass around the entire buffer along with an explicit length.

We could also have a special buffer type that keeps track of both a current slice but also has the full origin and length so the original buffer can be recovered. But that's probably best not to deal with in this PR.

Reply:
Yeah, right never mind didn't think about that :frowning_woman: